okio/src/main/java/okio/BufferCursor.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright (C) 2018 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package okio;
+
+/**
+ * Direct access to the underlying byte arrays in a buffer.
+ *
+ * Note that cursors are stateful and can be reused.
+ *
+ * It is an error to mutate a buffer that is currently being operated on with a cursor.
+ *
+ * Currently this cursor only supports reading. It is an error to mutate the fields of this cursor
+ * or the values of the data[] array.
+ *
+ * Access should be in acquire/release pairs. This guarded access may be used to detect bugs where
+ * cursors get out of sync with their underlying buffers.
+ *
+ * TODO(jwilson): nice, human-readable docs
+ * TODO(jwilson): support growing the buffer at the end
+ * TODO(jwilson): support acquiring for write (with seek() that gets writable segments)
+ */
+public final class BufferCursor {
+  private Buffer buffer;
+  public long size = -1L;
+
+  private Segment segment;
+  public long offset = -1L;
+  public byte[] data;
+  public int pos = -1;
+  public int limit = -1;
+
+  public void acquireReadOnly(Buffer buffer) {
```

Comment:
What if we want to start consuming at a specific offset?  The best I've been able to come up with the following, which wasn't very intuitive.  Might be more intuitive to have an `acquireReadOnly(buffer, offset)` that primed the `next()` to start at offset instead of 0.

```java
cursor.acquireReadOnly(buffer);
try {
  for (int length = cursor.seek(offset); toConsume > 0; length = cursor.next()) {
    length = (int) Math.min(length, toConsume);
    consume(cursor.data, cursor.pos, length);
    toConsume -= length;
  }
} finally {
  cursor.release();
}
```

Reply:
Could a cursor be used to simplify the `findSegmentAndOffset` labeled code in Buffer?

Reply:
I’m imagining some uses of cursors that don’t necessarily traverse the entire dataset. Perhaps a buffer contains sorted lines of text and we’re binary searching for a value. For this reason I think it’s easiest to keep acquire and seek separate.

We could definitely use cursors in `findSegmentAndOffset()`. Though I don’t think it would be profitable to do so. The core problem with that is that we’d either need to allocate a cursor on each use (inefficient) or forbid concurrent reads (breaks my effectively-immutable goal above).